### PR TITLE
core: improve locks in txpool

### PR DESCRIPTION
**Pull request by manav2401**
# Description

This PR cherry-picks https://github.com/maticnetwork/bor/pull/805 and https://github.com/maticnetwork/bor/pull/806 which improves locks on a map used by transaction pool. 

Thanks and shout out to @thogard785 for helping in identifying and fixing this issue. 

Quoting the description from #805 by @thogard785
> The filter method of txSortedMap is typically called by caller that holds the "m" RWLock object.
> 
> During reorgs, however, a Filter method of txList calls the filter method of txSortedMap but does not engage the lock.
> 
> This Filter method is called only by the txPool methods demoteUnexecutables and promoteExecutables, which are only called during reorgs (which explains the low frequency with which this crash occurs).

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

**Review by pratikspatil024 — APPROVED** [review]

**Review by Raneet10 — APPROVED** [review]

**Review by marcello33 — APPROVED** [review]

**Comment by kaber2** [comment]
Wow, you guys are truly the worst programmers I've seen in more than a decade. Conditional locking? Seriously? The proper solution is right before you eyes, but no, you merge garbage like this. OMG.

**Comment by JekaMas** [comment]
> Wow, you guys are truly the worst programmers I've seen in more than a decade. Conditional locking? Seriously? The proper solution is right before you eyes, but no, you merge garbage like this. OMG.

I don't think this is helpful. Would love to see your solution with proper benchmarks and profiling.
By far I suggest you not using offensive language. I don't think that such things are welcome in open source.

File: core/tx_list.go [hunk 1]
```diff
@@ -150,19 +151,38 @@ func (m *txSortedMap) Filter(filter func(*types.Transaction) bool) types.Transac
...
 	if len(removed) > 0 {
-		m.reheap()
+		m.reheap(false)
 	}
 	return removed
 }
 
-func (m *txSortedMap) reheap() {
-	*m.index = make([]uint64, 0, len(m.items))
+func (m *txSortedMap) reheap(withRlock bool) {
```
Comment (wjmelements) [inline core/tx_list.go:159]: This is not good practice. Instead you should have two methods, `Reheap` and `reheap`, where `Reheap` wraps `reheap` and takes the lock.

File: core/tx_list.go [hunk 2]
```diff
@@ -150,19 +151,38 @@ func (m *txSortedMap) Filter(filter func(*types.Transaction) bool) types.Transac
...
+	index := make(nonceHeap, 0, len(m.items))
+
+	if withRlock {
+		m.m.RLock()
+		log.Info("[DEBUG] Acquired lock over txpool map while performing reheap")
```
Comment (wjmelements) [inline core/tx_list.go:164]: Use log.Debug

File: core/tx_list.go [hunk 3]
```diff
@@ -521,12 +541,17 @@ func (l *txList) Filter(costLimit *uint256.Int, gasLimit uint64) (types.Transact
 				lowest = nonce
 			}
 		}
+
+		l.txs.m.Lock()
 		invalids = l.txs.filter(func(tx *types.Transaction) bool { return tx.Nonce() > lowest })
+		l.txs.m.Unlock()
```
Comment (wjmelements) [inline core/tx_list.go:547]: This should be `l.txs.Filter` instead of locking outside

**Pull request merged**